### ISSUE TITLE
Dropped sample in agent mode

### ISSUE THREAD

**Issue by lzfhust** [prometheus/prometheus#11589]
### What did you do?

1. we are running prometheus with victoriametrics as the remote storage. At the same time, prometheus runs with the local tsdb.
Now, we are trying to switch to the agent mode to reduce memory and cpu usage since we already have victoriametrics.
Many error logs happen: "Dropped sample for series that was not explicitly dropped via relabelling" ref=301997"
This means many samples are dropped ???

<img width="1390" alt="企业微信截图_e644b477-b2bf-4718-9fec-72d4dbc0cfa1" src="https://user-images.githubusercontent.com/9677371/202151335-428e091f-c5b3-4d60-a40b-f128bb01e163.png">

2.
2.1. I add a log here and the log is never printed, which means drop is not caused by relabel.
<img width="1044" alt="企业微信截图_6810dad6-92cb-44cd-9b09-4c802525b8d3" src="https://user-images.githubusercontent.com/9677371/202149851-a3a9d2aa-4633-4b9b-b7b6-870f344952f4.png">

2.2. Then I add a log here and the corresponding ref in "Dropped sample for series" log is printed, which means the ref is created and saved to wal file
<img width="843" alt="企业微信截图_c35d5a2e-e50e-4994-95aa-d90fef9e6d53" src="https://user-images.githubusercontent.com/9677371/202176914-f6256367-56b1-4781-9d97-91d81f9f98cc.png">

log detail:
ts=2022-11-16T11:59:32.261Z caller=db.go:726 level=info msg="tsdb agent append series ref " 308273=(MISSING)

ts=2022-11-16T11:59:47.491Z caller=dedupe.go:112 component=remote level=info remote_name=605597 url=http://10.0.228.33:8480/insert/-1/prometheus/api/v1/write msg="Dropped sample for series that was not explicitly dropped via relabelling" ref=308273
 
2.3. 
This problem happened in the first start up and before any checkpoints 
Now I don't know how to continue this problem....

### What did you expect to see?

No "Dropped sample" log happen

### What did you see instead? Under which circumstances?

errors as described

### Prometheus version

```text
prometheus, version 2.39.1 (branch: v2.39.1_recommend-test, revision: 3562f85a5b9f135262179ac7d6df9562e5b742a1)
  build user:       root@qsh4-sit-rec-reclambdaservice-dev-1
  build date:       20221105-06:58:37
  go version:       go1.19.2
  platform:         linux/amd64
```

### Prometheus configuration file

```yaml
no
```

### Alertmanager configuration file

```yaml
no
```

### Logs

```text
ts=2022-11-16T08:45:19.129Z caller=dedupe.go:112 component=remote level=info remote_name=605597 url=http://10.0.228.33:8480/insert/-1/prometheus/api/v1/write msg="Dropped sample for series that was not explicitly dropped via relabelling" ref=301997
ts=2022-11-16T08:45:19.129Z caller=dedupe.go:112 component=remote level=info remote_name=605597 url=http://10.0.228.33:8480/insert/-1/prometheus/api/v1/write msg="Dropped sample for series that was not explicitly dropped via relabelling" ref=301998
ts=2022-11-16T08:45:19.129Z caller=dedupe.go:112 component=remote level=info remote_name=605597 url=http://10.0.228.33:8480/insert/-1/prometheus/api/v1/write msg="Dropped sample for series that was not explicitly dropped via relabelling" ref=301999
ts=2022-11-16T08:45:19.129Z caller=dedupe.go:112 component=remote level=info remote_name=605597 url=http://10.0.228.33:8480/insert/-1/prometheus/api/v1/write msg="Dropped sample for series that was not explicitly dropped via relabelling" ref=323691
ts=2022-11-16T08:45:19.129Z caller=dedupe.go:112 component=remote level=info remote_name=605597 url=http://10.0.228.33:8480/insert/-1/prometheus/api/v1/write msg="Dropped sample for series that was not explicitly dropped via relabelling" ref=323692
ts=2022-11-16T08:45:19.129Z caller=dedupe.go:112 component=remote level=info remote_name=605597 url=http://10.0.228.33:8480/insert/-1/prometheus/api/v1/write msg="Dropped sample for series that was not explicitly dropped via relabelling" ref=323693
ts=2022-11-16T08:45:19.129Z caller=dedupe.go:112 component=remote level=info remote_name=605597 url=http://10.0.228.33:8480/insert/-1/prometheus/api/v1/write msg="Dropped sample for series that was not explicitly dropped via relabelling" ref=323694
ts=2022-11-16T08:45:19.129Z caller=dedupe.go:112 component=remote level=info remote_name=605597 url=http://10.0.228.33:8480/insert/-1/prometheus/api/v1/write msg="Dropped sample for series that was not explicitly dropped via relabelling" ref=323695
ts=2022-11-16T08:45:19.129Z caller=dedupe.go:112 component=remote level=info remote_name=605597 url=http://10.0.228.33:8480/insert/-1/prometheus/api/v1/write msg="Dropped sample for series that was not explicitly dropped via relabelling" ref=323696
ts=2022-11-16T08:45:19.129Z caller=dedupe.go:112 component=remote level=info remote_name=605597 url=http://10.0.228.33:8480/insert/-1/prometheus/api/v1/write msg="Dropped sample for series that was not explicitly dropped via relabelling" ref=323697
ts=2022-11-16T08:45:19.129Z caller=dedupe.go:112 component=remote level=info remote_name=605597 url=http://10.0.228.33:8480/insert/-1/prometheus/api/v1/write msg="Dropped sample for series that was not explicitly dropped via relabelling" ref=323698
ts=2022-11-16T08:45:19.129Z caller=dedupe.go:112 component=remote level=info remote_name=605597 url=http://10.0.228.33:8480/insert/-1/prometheus/api/v1/write msg="Dropped sample for series that was not explicitly dropped via relabelling" ref=323699
ts=2022-11-16T08:45:19.129Z caller=dedupe.go:112 component=remote level=info remote_name=605597 url=http://10.0.228.33:8480/insert/-1/prometheus/api/v1/write msg="Dropped sample for series that was not explicitly dropped via relabelling" ref=323700
ts=2022-11-16T08:45:19.129Z caller=dedupe.go:112 component=remote level=info remote_name=605597 url=http://10.0.228.33:8480/insert/-1/prometheus/api/v1/write msg="Dropped sample for series that was not explicitly dropped via relabelling" ref=323701
ts=2022-11-16T08:45:19.129Z caller=dedupe.go:112 component=remote level=info remote_name=605597 url=http://10.0.228.33:8480/insert/-1/prometheus/api/v1/write msg="Dropped sample for series that was not explicitly dropped via relabelling" ref=323702
ts=2022-11-16T08:45:19.129Z caller=dedupe.go:112 component=remote level=info remote_name=605597 url=http://10.0.228.33:8480/insert/-1/prometheus/api/v1/write msg="Dropped sample for series that was not explicitly dropped via relabelling" ref=323703
ts=2022-11-16T08:45:19.129Z caller=dedupe.go:112 component=remote level=info remote_name=605597 url=http://10.0.228.33:8480/insert/-1/prometheus/api/v1/write msg="Dropped sample for series that was not explicitly dropped via relabelling" ref=323704
ts=2022-11-16T08:45:19.129Z caller=dedupe.go:112 component=remote level=info remote_name=605597 url=http://10.0.228.33:8480/insert/-1/prometheus/api/v1/write msg="Dropped sample for series that was not explicitly dropped via relabelling" ref=323705
```

**Comment by roidelapluie** [comment]
cc @rfratto

**Comment by rfratto** [comment]
I'm guessing the series record is missing from the WAL but there are still sample records. 

Are you passing any flags to the agent mode to customize how frequency the truncate runs? I'm wondering if that's causing a bug where a series record gets prematurely.

**Comment by lzfhust** [comment]
> I'm guessing the series record is missing from the WAL but there are still sample records.
> 
> Are you passing any flags to the agent mode to customize how frequency the truncate runs? I'm wondering if that's causing a bug where a series record gets prematurely.

This is my start up command:
/bin/prometheus --config.file=/prometheus/prometheus-hash.yml --enable-feature=agent --web.listen-address=:9090 --web.enable-lifecycle --web.enable-admin-api --enable-feature=new-service-discovery-manager --enable-feature=extra-scrape-metrics --enable-feature=expand-external-labels --enable-feature=auto-gomaxprocs

**Comment by rfratto** [comment]
Can you try deleting the WAL used by the agent and seeing if this happens again?

**Comment by lzfhust** [comment]
> Can you try deleting the WAL used by the agent and seeing if this happens again?

@rfratto I have already tried this again, and this happened again.

**Comment by rfratto** [comment]
How quickly is it happening from process start time without an existing WAL in place?

**Comment by lzfhust** [comment]
> How quickly is it happening from process start time without an existing WAL in place?

It is a little strange. There are two prometheus agent to scrape metrics. One works well with very little drops, in the other one, it happend very quickly sometimes.

**Comment by lzfhust** [comment]
Now, it looks like in appending data to wal, the series data may be missing while the sample data is ok? But how can it be? In the commit, the series data is appended firstly and the sample is appended secondly....

**Comment by ptodev** [comment]
> This is my start up command: @rfratto /bin/prometheus --config.file=/prometheus/prometheus-hash.yml --enable-feature=agent --web.listen-address=:9090 --web.enable-lifecycle --web.enable-admin-api --enable-feature=new-service-discovery-manager --enable-feature=extra-scrape-metrics --enable-feature=expand-external-labels --enable-feature=auto-gomaxprocs

Hi @lzfhust, would you mind sharing the contents of the ```/prometheus/prometheus-hash.yml``` file please? I would like to try running a Prometheus with the same parameters to see if I can reproduce the issue.

**Comment by resec** [comment]
any update on this? we met the same issue with prometheus agent which unexpectedly drops metrics on remote writing.

**Comment by lzfhust** [comment]
> 



> > This is my start up command: @rfratto /bin/prometheus --config.file=/prometheus/prometheus-hash.yml --enable-feature=agent --web.listen-address=:9090 --web.enable-lifecycle --web.enable-admin-api --enable-feature=new-service-discovery-manager --enable-feature=extra-scrape-metrics --enable-feature=expand-external-labels --enable-feature=auto-gomaxprocs
> 
> Hi @lzfhust, would you mind sharing the contents of the `/prometheus/prometheus-hash.yml` file please? I would like to try running a Prometheus with the same parameters to see if I can reproduce the issue.

@ptodev Hi, my prometheus hash yaml file as below:
<img width="1108" alt="企业微信截图_740d1d4d-38e7-48fc-ad93-c290c3474aeb" src="https://user-images.githubusercontent.com/9677371/212894201-a0301cc8-a87e-46aa-9935-f5bef4d1bb5c.png">

**Comment by JuliaAlves** [comment]
Hey guys, any updates on this? I'm facing the same issues, mysterious and intermittent samples dropped 
cc. @ptodev

**Comment by ptodev** [comment]
I came across this error in a different place. I noticed it in a Grafana Agent, very close to startup time. It was so close to startup time that I believe there weren't even WAL checkpoint files at the time of the error. This is interesting, because it means that it might not be related to the [garbage collection](https://github.com/prometheus/prometheus/blob/v2.45.0/tsdb/wlog/watcher.go#L528) which the WAL watcher does. This garbage collection is done only when there are new checkpoints. So could it be that there is a segment file which has a sample before it has the series for that sample?

The fact that this problem is so hard to debug is also a problem in itself. The fact that the WAL gets truncated means there is no data for the ref IDs in the error message. Hence, I don't even know which series correspond to the ref IDs.

I'd suggest that we start tackling this issue by discussing how to debug such issues in an easier way:

* Should we have a config flag which tells the Agent to back up the WAL when it gets truncated, instead of deleting it? That way people could investigate this issue later on.
* We need to have a way to alert based on a metric if such an error happens. At the moment the metric for this error is also  incremented when a sample is dropped due to relabelling rules. Hence, it's not possible to tell if a sample got dropped due to this error or a relabelling rule. IIUC, a drop due to a relabelling rule is expected, and no alert should generally be generated. A drop due to this error is not expected, and can only be due to a bug (I can't think of any other reason). We need to be able to differentiate between those two situations in metrics.

**Comment by liguozhong** [comment]
any update on this?

**Comment by ptodev** [comment]
Below is a summary of my thoughts on this so far.

### A potential workaround

In theory this error could happen if the remote write is falling behind. If you are impacted by the error, I'd suggest going through this checklist:

1. Make sure that you see at least one of these three errors in your logs - if you don't see any of them, you must be affected by a different issue:

   - `Dropped sample for series that was not explicitly dropped via relabelling`
   - `Dropped exemplar for series that was not explicitly dropped via relabelling`
   - `Dropped histogram for series that was not explicitly dropped via relabelling`

2. Check if the `prometheus_remote_storage_enqueue_retries_total` metric is increasing around the time of the error. An increase in this metric means there is a buildup of data in the remote write queue.
3. Check your logs for a `Remote storage resharding` message, right around the time of the `dropped sample` error.
4. Check the `prometheus_remote_storage_shards_max` metric. Does it match the amount of shards reported by the `Remote storage resharding` log line? If it does, then there is a lot of data building up in the remote write queue. In that case, consider reconfiguring Prometheus to either use more shards or have a bigger shard capacity. There are configuration best practices on the [Remote Write Tuning](https://prometheus.io/docs/practices/remote_write/) page.

A remote write is falling behind is the only non-bug reason for this error that I can think of.

### What if the workaround doesn't work?

Please let us know on this issue with information such as:

- How quickly after startup this issue tends to occur.
- Please send us logs, especially with these messages around the time of the error:
   - `Remote storage resharding`
   - Search for logs containing `checkpoint` or `segment`, such as `Creating checkpoint` and `WAL checkpoint complete`.
- What is the `prometheus_remote_storage_shards_max` metric, and are the shards maxed out at the time of the error?
- Does the `prometheus_remote_storage_enqueue_retries_total` metric tend to increase at the time of the error?

The purpose of sending the above information is to determine if the error happens when remote write is overloaded, or if it happens during normal operation.

### Should we do dev work on Prometheus to regarding this?

I think so:

- In the metrics it would be good to differentiate the counter for "samples dropped due to relabeling" and "samples dropped due to a missing reference". We could either have separate metrics or the same metric but a different label. Some customers may want to have an alert on a "missing reference" metric, and potentially reconfigure Prometheus if it happens too often.
- If a "missing reference" error happens, we should have a way to tell if it's because the remote write is simply falling behind, or if it's due to a bug somewhere. Maybe we could spin off a separate goroutine which goes through the WAL segments and logs all the information it could find about this reference ID.
- Should we make the remote write more resilient when there's a buildup of work in the queues? For example, we could delay the garbage collection of series until we are confident that we have sent all the data for them?
- More logging about series getting garbage collected from the remote write queues?

**Comment by liguozhong** [comment]
> In theory this error could happen if the remote write is falling behind.

Thanks @ptodev for such a detailed report.
Is this a bug? If yes, is it possible to solve it through code repair? 

It doesn’t sound very reasonable to solve the bug by tuning parameters. 

I also use the agent mode to collect sla indicators on the collection line, hoping that the agent mode will not lose data.

**Comment by ptodev** [comment]
> Is this a bug? 

I'd say so. Even if a remote write falls behind, I wouldn't expect it to lose track of the reference ID of a series if the series still has samples in progress. I am currently trying to reproduce the error on my local machine and will also try to fix this somehow in the codebase. I will write another post next week to summarise my progress (hopefully there is progress 🤞 ).

**Comment by liguozhong** [comment]
thanks, all users of agent mode really like to see this problem solved

**Comment by ptodev** [comment]
Unfortunately I didn't have time to work on this this week, and I was unsuccessful with my attempts to replicate it. I used the Grafana Agent docker compose examples and only started an [Avalanche](https://github.com/grafana/agent/blob/v0.34.3/example/docker-compose/docker-compose.yaml#L54-L61), [Cortex](https://github.com/grafana/agent/blob/v0.34.3/example/docker-compose/docker-compose.yaml#L54-L61) and [Grafana](https://github.com/grafana/agent/blob/v0.34.3/example/docker-compose/docker-compose.yaml#L54-L61) + a local Prometheus on the side. I was hoping that generating new series every 2 min or so via the Avalanche would be enough to replicate this within several hours, but it wasn't. I'll try to pick this up at some point after my PTO next week, but can't commit to a time frame yet.

**Comment by jcleal13** [comment]
Hi @ptodev 

We're having the same problem here.

Seems insane, but I think it is related to samples limit exceeded. Look to both graphs


![image](https://github.com/prometheus/prometheus/assets/48995318/4ec44306-740a-445a-a7e3-68555053ebc4)

![image](https://github.com/prometheus/prometheus/assets/48995318/137a73ff-2dbc-4509-8db1-38569e1539f9)


Look at, when the samples limit start exceed, the dropped samples starts too.

What I know is that when a target from a job exceeded, only this target starts dropped sample.

But, why is this happen? Seems that the ref id is lost, but, if the timeseries wasn't scrape, why this happened?

Another situation. It would be good have a way to identify what the target is exceeding sample limit

**Comment by rfratto** [comment]
> Seems insane, but I think it is related to samples limit exceeded. Look to both graphs

This makes sense. The first time a new series is created in memory, is is queued to be written as a series record in the WAL. However, exceeding the sample limit causes the entire write to the WAL for that scrape (including the new series records) to be rolled back. However, in-memory series are still kept. 

This means that the next time a scrape happens with that series, it won't be considered new anymore, and the series record will never be written to the WAL.

This doesn't happen in Prometheus with a TSDB because it always writes series records to the WAL regardless of rollback.

**Comment by liguozhong** [comment]
> This means that the next time a scrape happens with that series

Will grafana/agent have similar problems？ Should we switch from prometheus-agent to grafana/agent to avoid data loss

**Comment by jcleal13** [comment]
> Edit: This can happen any time a scrape is rolled back, not just when the sample limit is exceeded.
@rfratto Is there a way to fix this without change the code? My only idea is increase the sample_limit, but this can be dangerous because I don't know how much I will need to increase.

I think run Prometheus agent completely stateless resolve because that way, i don't have the wal file and, by consequence I don't have to write in the wal file, make sense?

**Comment by diranged** [comment]
We're running Prometheus in stateless agent mode, and I believe we are seeing this problem pretty significantly. We do make heavy use of `sampleLimit` settings to manage our metric load.

**Comment by rfratto** [comment]
> @rfratto Is there a way to fix this without change the code? My only idea is increase the sample_limit, but this can be dangerous because I don't know how much I will need to increase.

Yes, any change to the config file which will lead to less frequent rollbacks will mitigate this issue. 

> I think run Prometheus agent completely stateless resolve because that way, i don't have the wal file and, by consequence I don't have to write in the wal file, make sense?

Yes, Prometheus Agent without the WAL is one way to resolve this issue, but that would require a more significant time investment to implement properly, and would leave users who rely on the WAL behind. 

We should try to fix this specific issue first IMO.

**Comment by jcleal13** [comment]
Yes, agree. Thank you so much @rfratto

**Comment by jcleal13** [comment]
> We're running Prometheus in stateless agent mode, and I believe we are seeing this problem pretty significantly. We do make heavy use of sampleLimit settings to manage our metric load.

We already are using a `sample_limit` so much large in my opinion, around `600000` in the jobs with the problem, but I will try increase more this number to try remediate the problem for now

**Comment by rfratto** [comment]
I've opened #12592 to address the rollback issue. Hopefully that fixes this issue for everyone, but I think we should leave this issue open for a while even after #12592 is merged to see if any reports for the same dropped sample issue come in.

**Comment by jcleal13** [comment]
Just to help others while the PR #12592 is not merged
We increased the `sampleLimit` in one Prometheus agent instance and in another not.
Look at the graphs that when the samples limit stopped exceeded, the dropped samples stopped too, in the another instance, both continued happening.
So, is that. For now, I think the only way to resolve, like @diranged said, is increase the sampleLimit in the jobs with problem. 
![image](https://github.com/prometheus/prometheus/assets/48995318/ce6951d8-dccf-467d-99b9-974ec40ee951)
![image](https://github.com/prometheus/prometheus/assets/48995318/2248a716-fa0e-4203-a71b-a11e616ded23)

**Comment by Vaibhav-1995** [comment]
Hi Team,

I'm also facing this issue in my setup, after adding below config to make some labels "key" to lowercase - for ex my label is **tag_Name** so want to modify it to **tag_name** and so on all labels starting with word "tag"
- action: labelmap
   regex: '${1}tag_([A-Za-z0-9_]+)'
   replacement: '${1}tag$2'

Getting below error log below - 
**ts=2024-04-16T11:28:36.798Z caller=dedupe.go:112 component=remote level=info remote_name=f08b29 url=https://xxx-grafana-mimir.XXXX.local/api/v1/push msg="Dropped sample for series that was not explicitly dropped via relabelling" ref=187571**

What changes are required to be done in prometheus chart to resolve this issue?

**Comment by hartfordfive** [comment]
Any updates on this and what the proposed resolution might be?   I'm also observing this error :
```
Dropped sample for series that was not explicitly dropped via relabelling"
```
although I'm not running Prometheus in agent mode and it's only occurring on a single instance of an HA pair of identical prometheus servers.  For reference, the prometheus version is 2.47.1.

**Comment by bboreham** [comment]
This issue should have been closed by #12592.  Please open a new issue with full details.
